File: packages/shared/dubwizard_shared/schemas/job.py
"""Job-related Pydantic schemas shared across components."""

from datetime import datetime
from typing import Optional
from pydantic import BaseModel, Field, field_validator

from dubwizard_shared.models.job_status import JobStatus


class JobCreate(BaseModel):
    """Schema for creating a new job."""
    target_language: str = Field(..., description="Target language for dubbing")
    voice_id: str = Field(..., description="ElevenLabs voice ID")
    source_language: str = Field(default="english", description="Source language of video")

    @field_validator("target_language")
    @classmethod
    def validate_target_language(cls, v):
        """Validate target language is supported."""
        supported = ["hindi"]
        if v.lower() not in supported:
            raise ValueError(f"Target language must be one of: {', '.join(supported)}")
        return v.lower()

    @field_validator("source_language")
    @classmethod
    def validate_source_language(cls, v):
        """Validate source language."""
        if v.lower() != "english":
            raise ValueError("Source language must be 'english' for MVP")
        return v.lower()


class JobResponse(BaseModel):
    """Schema for job response."""
    job_id: str
    status: JobStatus
    progress: int = Field(ge=0, le=100)
    source_language: str
    target_language: str
    voice_id: str
    video_duration_seconds: Optional[float] = None
    error_message: Optional[str] = None
    created_at: datetime
    updated_at: datetime
    completed_at: Optional[datetime] = None

    class Config:
        from_attributes = True


class JobStatusResponse(BaseModel):
    """Schema for job status response."""
    job_id: str
    status: JobStatus
    progress: int
    error_message: Optional[str] = None

    class Config:
        from_attributes = True


class JobDB(BaseModel):
    """Schema for job database model."""
    id: str
    status: JobStatus
    progress: int
    input_s3_key: str
    source_language: str
    target_language: str
    voice_id: str
    output_video_s3_key: Optional[str] = None
    source_subtitle_s3_key: Optional[str] = None
    target_subtitle_s3_key: Optional[str] = None
    video_duration_seconds: Optional[float] = None
    error_message: Optional[str] = None
    created_at: datetime
    updated_at: datetime
    completed_at: Optional[datetime] = None

    class Config:
        from_attributes = True

File: apps/web/src/types/job.ts
/**
 * Job-related TypeScript types matching backend schemas.
 */

export type JobStatus =
  | 'created'
  | 'uploading'
  | 'queued'
  | 'processing'
  | 'transcribing'
  | 'translating'
  | 'synthesizing'
  | 'processing_video'
  | 'done'
  | 'failed';

export interface Job {
  id: string;
  status: JobStatus;
  progress: number;
  input_s3_key: string;
  source_language: string;
  target_language: string;
  voice_id: string;
  output_video_s3_key?: string;
  source_subtitle_s3_key?: string;
  target_subtitle_s3_key?: string;
  video_duration_seconds?: number;
  error_message?: string;
  created_at: string;
  updated_at: string;
  completed_at?: string;
}

export interface CreateJobRequest {
  filename: string;
  file_size: number;
  content_type: string;
  target_language: string;
  voice_id: string;
  source_language?: string;
}

export interface CreateJobResponse {
  job_id: string;
  status: JobStatus;
  upload_url: string;
  s3_key: string;
}

export interface DownloadFile {
  url: string;
  filename: string;
  size_bytes: number;
  expires_in: number;
}

export interface SubtitleFiles {
  source: DownloadFile;
  target: DownloadFile;
}

export interface DownloadResponse {
  video: DownloadFile;
  subtitles: SubtitleFiles;
}

export interface VoiceOption {
  id: string;
  name: string;
  description: string;
  gender: 'male' | 'female';
}

// Predefined voice options for MVP (ElevenLabs voices)
export const VOICE_OPTIONS: VoiceOption[] = [
  {
    id: 'pNInz6obpgDQGcFmaJgB',
    name: 'Adam',
    description: 'Deep, warm male voice',
    gender: 'male',
  },
  {
    id: 'EXAVITQu4vr4xnSDxMaL',
    name: 'Bella',
    description: 'Soft, expressive female voice',
    gender: 'female',
  },
];

// Status display helpers
export const STATUS_LABELS: Record<JobStatus, string> = {
  created: 'Created',
  uploading: 'Uploading',
  queued: 'Queued',
  processing: 'Processing',
  transcribing: 'Transcribing Audio',
  translating: 'Translating',
  synthesizing: 'Generating Speech',
  processing_video: 'Processing Video',
  done: 'Complete',
  failed: 'Failed',
};

export const isJobComplete = (status: JobStatus): boolean => status === 'done';
export const isJobFailed = (status: JobStatus): boolean => status === 'failed';
export const isJobProcessing = (status: JobStatus): boolean =>
  !isJobComplete(status) && !isJobFailed(status);
